### PR TITLE
BTS-697 | GeoJSON changes for 3.10

### PR DESCRIPTION
**Pull request by neunhoef**
This is a PR to document the GeoJSON changes done for the 3.10 release.
This covers release notes as well as an overhaul of the description
of geo indexes and geo functions.

Potential changes in arangosearch are still missing at this stage.

### CODE REVIEW

File: 3.10/aql/functions-geo.md [hunk 1]
```diff
@@ -47,12 +47,29 @@ FOR doc IN doc // e.g. documents returned by a traversal
...
+```
+  FOR doc IN collectionname
+    FILTER GEO_CONTAINS(geoJson, doc.geo)
+    ...
+```
+
+can be **optimized** by an S2 based [geospatial index](../indexing-geo.html) 
+on the attribute `geo` of the collection `collectionname`, if `geoJson`
+evaluates to a valid `geojson` object, and the index has the `geoJson`
+flag set to `true`. Note that it has to be in this direction,
```
Comment (dcosta72) [inline 3.10/aql/functions-geo.md]: @neunhoef what do you mean by `direction`?

Comment (dcosta72) [inline 3.10/aql/functions-geo.md]: @neunhoef what do you mean by `direction`?

File: 3.10/aql/functions-geo.md [hunk 2]
```diff
@@ -45,14 +45,40 @@ FOR doc IN doc // e.g. documents returned by a traversal
 `GEO_CONTAINS(geoJsonA, geoJsonB) → bool`
 
 Checks whether the [GeoJSON object](../indexing-geo.html#geojson) `geoJsonA`
-fully contains `geoJsonB` (Every point in B is also in A). The object `geoJsonA` has to be of type 
-`Polygon` or `MultiPolygon`, other types are not supported because containment is ill defined. 
-This function can be **optimized** by a S2 based [geospatial index](../indexing-geo.html).
+fully contains `geoJsonB` (every point in B is also in A). The object `geoJsonA`
+has to be of type `Polygon` or `MultiPolygon`, other types are not supported
+because containment is ill-defined.
```
Comment (dcosta72) [inline 3.10/aql/functions-geo.md:50]: I'd rephrase this. 
Sth along this lines:
"The object 'geoJsonA' must conform to the currently supported types: 'Polygon' or 'MultiPolygon'. Support for other types will be introduced on future versions.

File: 3.10/aql/functions-geo.md [hunk 3]
```diff
@@ -45,14 +45,40 @@ FOR doc IN doc // e.g. documents returned by a traversal
...
 
 - **geoJsonA** (object): first GeoJSON object or coordinate array (in longitude, latitude order)
 - **geoJsonB** (object): second GeoJSON object or coordinate array (in longitude, latitude order)
-- returns **bool** (bool): true when every point in B is also contained in A, false otherwise
+- returns **bool** (bool): true if every point in B is also contained in A, false otherwise
 
+{% hint 'info' %}
+ArangoDB faithfully shows the same behavior as the underlying
```
Comment (dcosta72) [inline 3.10/aql/functions-geo.md:57]: ArangoDB follows and exposes the behaviour of the underlying S2 geometry library. As stated in the S2 documentation:

File: 3.10/aql/functions-geo.md [hunk 4]
```diff
@@ -45,14 +45,40 @@ FOR doc IN doc // e.g. documents returned by a traversal
...
+A query containing a `FILTER` expression of the form
+
+```js
+FOR doc IN collectionname
+  FILTER GEO_CONTAINS(geoJson, doc.geo)
+  ...
+```
+
+can be **optimized** by an S2-based [geospatial index](../indexing-geo.html)
+(on the attribute `geo` of the collection `collectionname` in this example), if
```
Comment (dcosta72) [inline 3.10/aql/functions-geo.md:77]: The conditional 'if' at the end reads difficult. Perhaps 
"...in this example), **only** if ...  "
After the conditions of IF, I'd add a line break.

File: 3.10/aql/functions-geo.md [hunk 5]
```diff
@@ -81,6 +107,29 @@ FOR doc IN collectionName
...
+A query containing a `FILTER` expression of the form
+
+```js
+FOR doc IN collectionname
+  FILTER GEO_DISTANCE(geoJson, doc.geo) <= limit
+  ...
+```
+
+can be **optimized** by an S2 based [geospatial index](../indexing-geo.html)
+(on the attribute `geo` of the collection `collectionname` in this example), if
```
Comment (dcosta72) [inline 3.10/aql/functions-geo.md:119]: Same comment as above re 'if'.

File: 3.10/indexing-index-basics.md [hunk 6]
```diff
@@ -369,13 +369,19 @@ The geo index stores two-dimensional coordinates. It can be created on either tw
 separate document attributes (latitude and longitude) or a single array attribute that
 contains both latitude and longitude. Latitude and longitude must be numeric values.
 
+Furthermore, a geo index can also index GeoJSON objects. GeoJSON is a
```
Comment (dcosta72) [inline 3.10/indexing-index-basics.md:372]: "... a geo index can also index starndard GeoJSON objects" (and the link on GeoJSON).
"GeoJSON uses JSON syntax to describe geomet..."

File: 3.10/indexing-which-index.md [hunk 7]
```diff
@@ -128,6 +128,11 @@ different usage scenarios:
   and a SORT or FILTER statement is used in conjunction with the distance
   function.
 
+  Furthermore, the geo index and AQL can work with GeoJSON objects. GeoJSON is a
```
Comment (dcosta72) [inline 3.10/indexing-which-index.md:131]: same comment as above

File: 3.10/release-notes-new-features310.md [hunk 8]
```diff
@@ -29,7 +29,30 @@ no move operations have been scheduled if they are not necessary.
 AQL
 ---
 
+### GeoJSON changes
```
Comment (dcosta72) [inline 3.10/release-notes-new-features310.md:32]: This is the controversial point. As @neunhoef mentioned, we need to carefully craft this message. Perhaps spinning it around.

"The 3.10 release of ArangoDB conforms to the standards specified in GeoJSON objects... and models...
This diverges from the previous implementation in two fundamental ways:
1.   ...
2. ....

For existing users that do not wish to rebuild their indices and continue using the previous behaviour, a flag 'legacyPolygon' [...] has been introduced to guarantee backwards compatibility.
For existing users that wish to take advantage of the new standard behaviour,, geo indexes need to be dropped and.... [...]

See .[...] upgrading to version 3.10 or later.
....
Additionally, the reported issues that where occasionally producing wrong results in geo queries when using geo indexes have been fixed. " (and perhaps link to the detail of what's been fixed?)

File: 3.10/release-notes-upgrading-changes310.md [hunk 9]
```diff
@@ -34,6 +34,29 @@ The fulltext index type is now deprecated in favor of [ArangoSearch](arangosearc
...
+correctly) in the newer versions.
+
+Legacy geo indexes will continue to work and continue to produce the
+same results as in earlier versions, since they will have the flag
+`legacyPolygons` implicitly set to `true`.
+
+Newly created indexes will have `legacyPolygons` set to `false` by default
+and thus enable the new polygon parsing.
+
+Note that linear rings that are boundaries of polygons
```
Comment (dcosta72) [inline 3.10/release-notes-upgrading-changes310.md:51]: I'd completely break from explaining what we were doing before.
"Note that, from ArangoDB 3.10, **as per the standard**, linear rings [...] are not normalized automatically. The 'interior' of a polygon strictly conforms to the GeoJSON standard: ..."
"Note that this differs from the < 3.10 interpretation of GeoJSON polygons (and link to the old documentation)"

**Review by dcosta72 — CHANGES_REQUESTED** [review]
@neunhoef added a few comments.

File: 3.10/aql/functions-geo.md (hunk 1 above)
Comment (neunhoef) [reply 3.10/aql/functions-geo.md]: I meant "order of arguments", clarified.

File: 3.10/aql/functions-geo.md (hunk 2 above)
Comment (neunhoef) [reply 3.10/aql/functions-geo.md:50]: We should not promise the impossible. Containment of points in lines is not numerically stable (which was meant by this "ill-defined" word. All other GeoJSON objects are zero- or one-dimensional and run into these numerical problems w.r.t. containment. Therefore we would not be wise to promise this. I will change the wording to explain this.

File: 3.10/aql/functions-geo.md (hunk 3 above)
Comment (neunhoef) [reply 3.10/aql/functions-geo.md:57]: Done.

File: 3.10/aql/functions-geo.md (hunk 5 above)
Comment (neunhoef) [reply 3.10/aql/functions-geo.md:119]: Done.

File: 3.10/indexing-index-basics.md (hunk 6 above)
Comment (neunhoef) [reply 3.10/indexing-index-basics.md:372]: Done.

File: 3.10/release-notes-new-features310.md (hunk 8 above)
Comment (neunhoef) [reply 3.10/release-notes-new-features310.md:32]: Changed.

File: 3.10/release-notes-upgrading-changes310.md (hunk 9 above)
Comment (neunhoef) [reply 3.10/release-notes-upgrading-changes310.md:51]: Done.

File: 3.10/indexing-which-index.md (hunk 7 above)
Comment (neunhoef) [reply 3.10/indexing-which-index.md]: Done.

**Review by dcosta72 — APPROVED** [review]
looks good

**Review by Simran-B — CHANGES_REQUESTED** [review]

File: 3.10/indexing-geo.md [hunk 10]
```diff
@@ -38,17 +41,87 @@ This creates the index on all documents and uses _geometry_ as the attributed
...
 to the format described in
 [RFC 7946 Position](https://tools.ietf.org/html/rfc7946#section-3.1.1){:target="_blank"}.
 
-All documents, which do not have the attribute path or have a non-conform
+All documents, that do not have the attribute path or have a non-conforming
 value in it, are excluded from the index.
 
-A geo index is implicitly sparse, and there is no way to control its sparsity.
+A geo index is implicitly sparse, and there is no way to control its
+sparseness.
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply 3.10/indexing-geo.md:52]: I have no strong feelings, but consistency is probably a good idea. Maybe we should ask a native speaker.

File: 3.10/indexing-geo.md [hunk 11]
```diff
@@ -206,6 +317,79 @@ Supported geometry object types are:
 - Polygon
 - MultiPolygon
 
+### GeoJSON interpretation
+
+Note the following technical detail about GeoJSON: The
+[GeoJSON standard, Section 3.1.1 Position](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.1){:target="_blank"}
+prescribes that lines are **cartesian lines in cylindrical coordinates**
+(longitude/latitude). However, this definition is awkward in practice,
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply 3.10/indexing-geo.md:325]: No strong feelings, either, but I intentionally wrote "awkward" instead of "problematic" to avoid being negative. But honestly, I do not really care.

File: 3.10/indexing-geo.md [hunk 12]
```diff
@@ -206,6 +317,79 @@ Supported geometry object types are:
...
+since such lines are not geodesic on the surface of the Earth.
+Furthermore, the best available algorithms for geospatial computations on Earth
+typically use geodesic lines as the boundaries of polygons on Earth.
+
+Therefore, ArangoDB uses the **syntax of the GeoJSON** standard,
+but then interpret lines (and boundaries of polygons) as
+**geodesic lines (pieces of great circles) on Earth**. This is a
+violation of the GeoJSON standard, but one which is sensible in practice.
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply 3.10/indexing-geo.md:333]: I have no opinion on this, but also see no reason to avoid "sensible".

File: 3.10/indexing-geo.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (ansoboleva) [reply 3.10/indexing-geo.md:52]: @Simran-B Both variants are used interchangeably, with sparsity having a more frequent usage. Please stick to sparsity and make sure it's consistent throughout the doc.

File: 3.10/indexing-geo.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (ansoboleva) [reply 3.10/indexing-geo.md:325]: @Simran-B inconvenient or cumbersome, but not problematic

File: 3.10/indexing-geo.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (ansoboleva) [reply 3.10/indexing-geo.md:333]: @Simran-B "a violation of the standard, but a practical one" or "serving/done for a practical purpose?"

File: 3.10/indexing-geo.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (Simran-B) [reply 3.10/indexing-geo.md:52]: There are no other occurrences of "sparseness" in the docs and main repo.
